README
EDO is a tool for selecting Ensemble-Diversity Optimised hairpins used as
barcodes in the context of high throughput synthesis pipelines, like Cloud Lab
(see http://daslab.stanford.edu/das_research.html)

Partition functions are computed using the ViennaRNA package.
http://www.tbi.univie.ac.at/RNA/

The distributed auction algorithm is largely based on ideas presented in
  Zavlanos, M.M.; Spesivtsev, L.; Pappas, G.J., 
  "A distributed auction algorithm for the assignment problem,"
  47th IEEE Conference on Decision and Control, 2008. CDC 2008.
  pp.1212-1217, 9-11 Dec. 2008
which can be found online at 
  http://dx.doi.org/10.1109/CDC.2008.4739098
  or http://repository.upenn.edu/grasp_papers/33/


Build the usual way:

  ./configure
  make

and then try

  src/edo -v < data/r89f.list >data/r89f.seqs
  src/edo --eval < data/r89f.seqs


ChangeLog
2014-10-27  Fernando Portela  <nando8888@gmail.com>

  * implemented the notion of 'lab settings'

2014-10-25  Fernando Portela  <nando8888@gmail.com>

  * code refactored, folding routines in edo_fold.cpp
  * switch from manual build to autotools

src/edo_cmdl.h
/** @file edo_cmdl.h
 *  @brief The header file for the command line option parser
 *  generated by GNU Gengetopt version 2.22.6
 *  http://www.gnu.org/software/gengetopt.
 *  DO NOT modify this file, since it can be overwritten
 *  @author GNU Gengetopt by Lorenzo Bettini */

#ifndef EDO_CMDL_H
#define EDO_CMDL_H

/* If we use autoconf.  */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdio.h> /* for FILE */

#ifdef __cplusplus
extern "C" {
#endif /* __cplusplus */

#ifndef EDO_CMDLINE_PARSER_PACKAGE
/** @brief the program name (used for printing errors) */
#define EDO_CMDLINE_PARSER_PACKAGE PACKAGE
#endif

#ifndef EDO_CMDLINE_PARSER_PACKAGE_NAME
/** @brief the complete program name (used for help and version) */
#ifdef PACKAGE_NAME
#define EDO_CMDLINE_PARSER_PACKAGE_NAME PACKAGE_NAME
#else
#define EDO_CMDLINE_PARSER_PACKAGE_NAME PACKAGE
#endif
#endif

#ifndef EDO_CMDLINE_PARSER_VERSION
/** @brief the program version */
#define EDO_CMDLINE_PARSER_VERSION VERSION
#endif

/** @brief Where the command line options are stored */
struct edo_args_info
{
  const char *help_help; /**< @brief Print help and exit help description.  */
  const char *version_help; /**< @brief Print version and exit help description.  */
  int eval_flag;	/**< @brief Evaluation mode.
 (default=off).  */
  const char *eval_help; /**< @brief Evaluation mode.
 help description.  */
  char * lab_settings_arg;	/**< @brief Read lab settings from FILE, instead of using the default settings (Cloud Lab).
.  */
  char * lab_settings_orig;	/**< @brief Read lab settings from FILE, instead of using the default settings (Cloud Lab).
 original value given at command line.  */
  const char *lab_settings_help; /**< @brief Read lab settings from FILE, instead of using the default settings (Cloud Lab).
 help description.  */
  char * paramFile_arg;	/**< @brief Read energy parameters from FILE, instead of using the default parameter set.
.  */
  char * paramFile_orig;	/**< @brief Read energy parameters from FILE, instead of using the default parameter set.
 original value given at command line.  */
  const char *paramFile_help; /**< @brief Read energy parameters from FILE, instead of using the default parameter set.
 help description.  */
  int verbose_flag;	/**< @brief Verbose output
 (default=off).  */
  const char *verbose_help; /**< @brief Verbose output
 help description.  */
  
  unsigned int help_given ;	/**< @brief Whether help was given.  */
  unsigned int version_given ;	/**< @brief Whether version was given.  */
  unsigned int eval_given ;	/**< @brief Whether eval was given.  */
  unsigned int lab_settings_given ;	/**< @brief Whether lab-settings was given.  */
  unsigned int paramFile_given ;	/**< @brief Whether paramFile was given.  */
  unsigned int verbose_given ;	/**< @brief Whether verbose was given.  */

} ;

/** @brief The additional parameters to pass to parser functions */
struct edo_cmdline_parser_params
{
  int override; /**< @brief whether to override possibly already present options (default 0) */
  int initialize; /**< @brief whether to initialize the option structure edo_args_info (default 1) */
  int check_required; /**< @brief whether to check that all required options were provided (default 1) */
  int check_ambiguity; /**< @brief whether to check for options already specified in the option structure edo_args_info (default 0) */
  int print_errors; /**< @brief whether getopt_long should print an error message for a bad option (default 1) */
} ;

/** @brief the purpose string of the program */
extern const char *edo_args_info_purpose;
/** @brief the usage string of the program */
extern const char *edo_args_info_usage;
/** @brief the description string of the program */
extern const char *edo_args_info_description;
/** @brief all the lines making the help output */
extern const char *edo_args_info_help[];

/**
 * The command line parser
 * @param argc the number of command line options
 * @param argv the command line options
 * @param args_info the structure where option information will be stored
 * @return 0 if everything went fine, NON 0 if an error took place
 */
int edo_cmdline_parser (int argc, char **argv,
  struct edo_args_info *args_info);

/**
 * The command line parser (version with additional parameters - deprecated)
 * @param argc the number of command line options
 * @param argv the command line options
 * @param args_info the structure where option information will be stored
 * @param override whether to override possibly already present options
 * @param initialize whether to initialize the option structure my_args_info
 * @param check_required whether to check that all required options were provided
 * @return 0 if everything went fine, NON 0 if an error took place
 * @deprecated use edo_cmdline_parser_ext() instead
 */
int edo_cmdline_parser2 (int argc, char **argv,
  struct edo_args_info *args_info,
  int override, int initialize, int check_required);

/**
 * The command line parser (version with additional parameters)
 * @param argc the number of command line options
 * @param argv the command line options
 * @param args_info the structure where option information will be stored
 * @param params additional parameters for the parser
 * @return 0 if everything went fine, NON 0 if an error took place
 */
int edo_cmdline_parser_ext (int argc, char **argv,
  struct edo_args_info *args_info,
  struct edo_cmdline_parser_params *params);

/**
 * Save the contents of the option struct into an already open FILE stream.
 * @param outfile the stream where to dump options
 * @param args_info the option struct to dump
 * @return 0 if everything went fine, NON 0 if an error took place
 */
int edo_cmdline_parser_dump(FILE *outfile,
  struct edo_args_info *args_info);

/**
 * Save the contents of the option struct into a (text) file.
 * This file can be read by the config file parser (if generated by gengetopt)
 * @param filename the file where to save
 * @param args_info the option struct to save
 * @return 0 if everything went fine, NON 0 if an error took place
 */
int edo_cmdline_parser_file_save(const char *filename,
  struct edo_args_info *args_info);

/**
 * Print the help
 */
void edo_cmdline_parser_print_help(void);
/**
 * Print the version
 */
void edo_cmdline_parser_print_version(void);

/**
 * Initializes all the fields a edo_cmdline_parser_params structure 
 * to their default values
 * @param params the structure to initialize
 */
void edo_cmdline_parser_params_init(struct edo_cmdline_parser_params *params);

/**
 * Allocates dynamically a edo_cmdline_parser_params structure and initializes
 * all its fields to their default values
 * @return the created and initialized edo_cmdline_parser_params structure
 */
struct edo_cmdline_parser_params *edo_cmdline_parser_params_create(void);

/**
 * Initializes the passed edo_args_info structure's fields
 * (also set default values for options that have a default)
 * @param args_info the structure to initialize
 */
void edo_cmdline_parser_init (struct edo_args_info *args_info);
/**
 * Deallocates the string fields of the edo_args_info structure
 * (but does not deallocate the structure itself)
 * @param args_info the structure to deallocate
 */
void edo_cmdline_parser_free (struct edo_args_info *args_info);

/**
 * Checks that all the required options were specified
 * @param args_info the structure to check
 * @param prog_name the name of the program that will be used to print
 *   possible errors
 * @return
 */
int edo_cmdline_parser_required (struct edo_args_info *args_info,
  const char *prog_name);


#ifdef __cplusplus
}
#endif /* __cplusplus */
#endif /* EDO_CMDL_H */


src/edo_fold.h
// Copyright (C) 2014 Fernando Portela <nando8888@gmail.com>
//  
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software 
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.



typedef struct _lab_config {
    int	  bclen;
    const char*  tail5p_ss;
    const char*  hairpin_ss;
    const char*  tail3p_ss;
    const char*  tail5p_nts;
    const char*  hairpin_nts;
    const char*  tail3p_nts;
    // forbidden subsequences
    const char** forbidden;
    // various precomputed offsets
    int   hp5p, hp3p;
    int   tl5p, tl3p, hp_tl;
    // scores
    double s_max, s_scale;
} lab_config;


extern int verbose;
extern lab_config* cnf;
extern int num_scored;

bool load_config( char* filename );
void load_fold_params( char* filename );
bool is_legal( char* seq );
double score_seq( int s, char* seq, int ip );


src/edo_fold.cpp
// Copyright (C) 2014 Fernando Portela <nando8888@gmail.com>
//  
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software 
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


#include <string.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

extern "C" {
#include "libconfig.h"

#include "RNAstruct.h"
#include "fold_vars.h"
#include "fold.h"
#include "params.h"
#include "part_func.h"
#include "utils.h"
#include "convert_epars.h"
#include "read_epars.h"
#include "MEA.h"
}

#include "edo_fold.h"


#define BC_LEN        7

#define TAIL5P_SS     ".."
#define HAIRPIN_SS    "(((((((....)))))))."
#define TAIL3P_SS     "...................."

#define TAIL5P_NTS    "GG"
#define HAIRPIN_NTS   ".......UUCG.......A"
#define TAIL3P_NTS    "AAAGAAACAACAACAACAAC"

static const char* daslab_forbidden[] = { "CCCC", "GGGG", NULL };

static lab_config daslab_cnf = {
    BC_LEN,
    TAIL5P_SS,
    HAIRPIN_SS,
    TAIL3P_SS,
    TAIL5P_NTS,
    HAIRPIN_NTS,
    TAIL3P_NTS,
    daslab_forbidden,
    39, 0, 11, 2,
    10., 1.
};

lab_config* cnf = &daslab_cnf;



bool load_config( char* filename )
{
    bool ok = true;
    config_t cfg;
    config_init( &cfg );
    if( config_read_file( &cfg, filename ) == CONFIG_TRUE ) {
        lab_config* lab = (lab_config*) calloc( 1, sizeof( lab_config ) );
        const char* buf;
        if( ok && config_lookup_string( &cfg, "settings.secstr.tail5p",
                                        &buf ) == CONFIG_TRUE ) {
            lab->tail5p_ss = strdup( buf );
        } else {
            ok = false;
        }
        if( ok && config_lookup_string( &cfg, "settings.secstr.hairpin",
                                        &buf ) == CONFIG_TRUE ) {
            lab->hairpin_ss = strdup( buf );
        } else {
            ok = false;
        }
        if( ok && config_lookup_string( &cfg, "settings.secstr.tail3p",
                                        &buf ) == CONFIG_TRUE ) {
            lab->tail3p_ss = strdup( buf );
        } else {
            ok = false;
        }
        if( ok && config_lookup_string( &cfg, "settings.sequences.tail5p",
                                        &buf ) == CONFIG_TRUE ) {
            lab->tail5p_nts = strdup( buf );
        } else {
            ok = false;
        }
        if( ok && config_lookup_string( &cfg, "settings.sequences.hairpin",
                                        &buf ) == CONFIG_TRUE ) {
            lab->hairpin_nts = strdup( buf );
        } else {
            ok = false;
        }
        if( ok && config_lookup_string( &cfg, "settings.sequences.tail3p",
                                        &buf ) == CONFIG_TRUE ) {
            lab->tail3p_nts = strdup( buf );
        } else {
            ok = false;
        }
        if( ok ) {
            // consistency checks
            ok &= strlen( lab->tail5p_ss ) == strlen( lab->tail5p_nts );
            ok &= strlen( lab->hairpin_ss ) == strlen( lab->hairpin_nts );
            ok &= strlen( lab->tail3p_ss ) == strlen( lab->tail3p_nts );
        }
        if( ok ) {
            char key[128];
            int i;
            for( i = 0; /* forever */; i++ ) {
                sprintf( key, "settings.forbidden.[%d]", i );
                if( config_lookup_string( &cfg, key, &buf ) != CONFIG_TRUE) break;
                lab->forbidden = (const char**) realloc( lab->forbidden,
                                                         (i+1) * sizeof( char* ) );
                lab->forbidden[i] = strdup( buf );
            }
            if( lab->forbidden ) {
                lab->forbidden = (const char**) realloc( lab->forbidden,
                                                         (i+1) * sizeof( char* ) );
                lab->forbidden[i] = NULL;
            }
        }
        if( ok ) {
            // populate the other fields in the lab settings
            lab->hp5p = strcspn( lab->hairpin_nts, "." );
            lab->bclen = strspn( lab->hairpin_nts + lab->hp5p, "." );
            int j = lab->hp5p + lab->bclen;
            lab->hp3p = j + strcspn( lab->hairpin_nts + j, "." );
            lab->tl5p = strlen( lab->tail5p_ss );
            lab->tl3p = strlen( lab->tail3p_ss );
            lab->hp_tl = strlen( lab->hairpin_ss ) + lab->tl5p + lab->tl3p;
            
            lab->s_max = round( strlen( lab->hairpin_ss ) + .9 ) * 0.5;
            lab->s_scale = 10.0 / lab->s_max;

            if( verbose ) {
                fprintf( stderr, "%s / / %s / / %s\n", lab->tail5p_nts,
                                                       lab->hairpin_nts,
                                                       lab->tail3p_nts );
                fprintf( stderr, "%s / / %s / / %s\n", lab->tail5p_ss,
                                                       lab->hairpin_ss,
                                                       lab->tail3p_ss );
                fprintf( stderr, "[%d,%d,%d,%d]-[%4.1f,%6.4f]\n",
                                 lab->hp_tl, lab->hp5p, lab->hp3p, lab->tl5p,
                                 lab->s_max, lab->s_scale );
                fprintf( stderr, "Configuration successfully loaded from %s\n",
                                 filename );
            }

            // activate
            cnf = lab;

        } else {
            free( lab );
        }
        
    } else {
        ok = false;
        fprintf( stderr, "%s:%d: %s\n", filename, config_error_line( &cfg ),
                                                  config_error_text( &cfg ) );
    }

    config_destroy( &cfg );
    return ok;
}


void load_fold_params( char* filename ) {
    read_parameter_file( filename );
}


bool is_legal( char* seq )
{
#if 0
    if( strstr( seq, "AAAAA" ) ) return false;
    if( strstr( seq, "CCCC" ) ) return false;
    if( strstr( seq, "GGGG" ) ) return false;
#endif

    if( cnf->forbidden ) {
        const char** p = cnf->forbidden;
        while( *p ) {
            if( strstr( seq, *p ) ) return false;
            p++;
        }
    }

    return true;
}


double score_seq( int s, char* seq, int ip )
{
    double score = 0.0;
    if( !is_legal( seq ) ) return score;

    int length = strlen( seq );
    int* ix = get_iindx( length );

    double betaScale = 1.0;
    double kT = ( betaScale*( ( temperature+K0 )*GASCONST ) )/1000.; /* in Kcal */
    model_detailsT md;
    set_model_details( &md );

    char* secstr = strdup( seq );
    secstr[0] = 0;
    fold_constrained = 0;
    paramT* params = get_scaled_parameters( temperature, md );
    double min_en = fold_par( seq, secstr, params, fold_constrained, 0 );
    if( strncmp( secstr + ip,
                 cnf->hairpin_ss, strlen( cnf->hairpin_ss ) ) != 0
        || strcmp( secstr + length - strlen( cnf->tail3p_ss ),
                   cnf->tail3p_ss ) != 0 ) {
        free( params );
        free( secstr );
        free( ix );
        return score;
    }
    #pragma omp atomic update
    num_scored++;

    double pf_scale = exp( -( 1.07*min_en )/kT/length );
    pf_paramT* pf_params = get_boltzmann_factors( temperature, betaScale, md, pf_scale );
    // Either patch fold_vars.h by inserting following at line 166
    //
    // #ifdef _OPENMP
    // #pragma omp threadprivate(iindx)
    // #endif
    //
    // or uncomment this pragma below
    //
    // #pragma omp critical(pf_fold)
    double e = pf_fold_par( seq, NULL, pf_params, 1, fold_constrained, 0 );
    FLT_OR_DBL* ppm = export_bppm();

#define pr_ij(i,j) (i == j? 0.0 : (i < j ? ppm[ix[i]-j] : ppm[ix[j]-i]))

    score = cnf->s_max;
    int i, o;
    for( i = 1; i <= length; i++ ) {
        for( o = 1; o <= strlen( cnf->hairpin_ss ); o++ ) {
            int j = ip + o;
            double v = pr_ij( i, j );
            score -= v * ( 1.0 - v );
        }
    }
    score *= cnf->s_scale;

    free( pf_params );
    free( params );
    free( secstr );
    free( ix );
    return score;
}


src/edo.cpp
// Copyright (C) 2014 Fernando Portela <nando8888@gmail.com>
//  
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software 
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


#include <stdlib.h>
#include <time.h>
#include <string.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <limits.h>
#include <unistd.h>
#include <omp.h>

#include <map>
#include <set>
#include <vector>
#include <list>
#include <queue>
#include <string>

#include "edo_fold.h"
#include "edo_cmdl.h"

#define BASES "AUGC"
#define SEPARATORS "_-+='*"


int verbose = 0;

float epsilon = 0.01;

int num_seq;
int num_bc;
int num_loc;
int t = 0;

float**        prices[2];    // [][num_seq][num_bc]
int**          bidders[2];   // [][num_seq][num_bc]
long*          alpha[2];     // [][num_seq]
long long*     locations[2]; // [][num_seq]

typedef std::map<long long, float> val_map;
typedef val_map::iterator          val_map_it;
typedef std::vector<val_map>       val_map_array;

char**         sequences = NULL;    //   [num_seq]
int*           hip = NULL;          //   [num_seq]
               // as in Hairpin Insertion Point
val_map_array  values;

char**         seqs = NULL;
int            nseqs = 0;

std::set<long> tabu;

int            num_scored = 0;




void init( void )
{
    int s;
    for( s = 0, num_bc = 1; s < cnf->bclen; s++ ) num_bc *= 4;
    for( s = 0, num_loc = 1; s < cnf->bclen; s++ ) num_loc *= 6;

    int t;
    for( t = 0; t < 2; t++ ) {
        prices[t] = ( float** ) calloc( num_seq, sizeof( float* ) );
        for( s = 0; s < num_seq; s++ )
            prices[t][s] = ( float* ) calloc( num_bc, sizeof( float ) );
        bidders[t] = ( int** ) calloc( num_seq, sizeof( int* ) );
        for( s = 0; s < num_seq; s++ )
            bidders[t][s] = ( int* ) calloc( num_bc, sizeof( int ) );
        alpha[t] = ( long* ) calloc( num_seq, sizeof( long ) );
        locations[t] = ( long long* ) calloc( num_seq, sizeof( long long ) );
    }

    sequences = (char**) calloc( num_seq, sizeof(char*) );
    values.resize( num_seq );

    srand48( time( NULL ) );
}

#if 0
bool neighbor( int i, int j )
{
    int k;
    unsigned int a = locations[t][i];
    unsigned int b = locations[t][j];
    int d = 0;
    for( k = 0; k < 7; k++ ) {
        if( ( a & 0xF ) != ( b & 0xF ) ) {
            d++;
            if( d > 1 ) return false;
        }
        a >>= 4;
        b >>= 4;
    }
    return true;
}
#endif


int int_urn( int from, int to )
{
    return ( ( (int) (drand48()*(to-from+1)) ) + from );
}
 
// Internal bitmap representation
//
// MSB  LSB  nucleobase    MSB    LSB   5'  3'
//    00        C             0011        CG
//    01        U             1100        GC
//    10        A             1001        AU
//    11        G             0110        UA
//                            0111        UG
//                            1101        GU
//
//
// MSB  (64 bits => up to 16 base pairs) LSB
//
//                ---++---++---++---++--  5'
//              ... -dddd-cccc-bbbb-aaaa
//                  ---++---++---++---++  3'
//

char bases[] = "CUAG";
long long pmap[] = { 0x03, 0x0C, 0x09, 0x06, 0x07, 0x0D };


long location_to_barcode( long long loc )
{
    int k;
    long c = 0;
    for( k = 0; k < cnf->bclen; k++ ) {
        c <<= 2;
        c |= ( loc & 0x03 );
        loc >>= 4;
    }
    return c;
}

void location_to_bcseq( long long loc, char* s1, char* s2 )
{
    int k;
    for( k = 0; k < cnf->bclen; k++ ) {
        s1[k  ] = bases[( loc & 0x0C )>>2];
        s2[6-k] = bases[loc & 0x03];
        loc >>= 4;
    }
}

long bcseq_to_barcode( char* seg )
{
    int k;
    long c = 0;
    for( k = 0; k < cnf->bclen; k++ ) {
        c <<= 2;
        c |= seg[k]=='C'? 0 : ( seg[k]=='U'? 1 : ( seg[k]=='A'? 2 : 3 ) );
    }
    return c;
}


bool is_tabu( char* seg )
{
    long c = bcseq_to_barcode( seg );
    return tabu.find( c ) != tabu.end();
}

void compute_value( int i, long long a )
{
    int len = 1 + strlen( sequences[i] ) + cnf->hp_tl;
    char* seq = (char*) malloc( len );
    sprintf( seq, "%s%.*s%s%s%s", cnf->tail5p_nts,
                                  hip[i], sequences[i],
                                  cnf->hairpin_nts, 
                                  sequences[i] + hip[i],
                                  cnf->tail3p_nts );
    location_to_bcseq( a, seq + cnf->tl5p + hip[i] + cnf->hp5p,
                          seq + cnf->tl5p + hip[i] + cnf->hp3p );
    values[i][a] = is_tabu( seq + cnf->tl5p + hip[i] + cnf->hp3p ) ?
                       0.0 : score_seq( i, seq, cnf->tl5p + hip[i] );
    free( seq );
}

void update_local_values( int i )
{
    int k;
    #pragma omp parallel for
    for( k = 0; k < cnf->bclen; k++ ) {
        int j;
        long long mask = 0x0F << ( k*4 );
        for( j = 0; j < 6; j++ ) {
            long long a = ( locations[t][i] & ~mask ) | ( pmap[j] << ( k*4 ) );
            if( values[i].find( a ) != values[i].end() ) continue;
            compute_value( i, a );
        }
    }
}


void do_round( int r, int* num_swap, float* delta )
{
    int i, cnt = 0;
    int this_num_swap;
    if( !num_swap ) num_swap = &this_num_swap;
    (*num_swap) = 0;
    float this_delta;
    if( !delta ) delta = &this_delta;
    (*delta) = 0.0;

    #pragma omp parallel for schedule(dynamic)
    for( i = 0; i < num_seq; i++ ) {
        int j;
        long k;
        int next_t = (t+1)%2;

        if( verbose )
        #pragma omp critical(console)
        {
            fprintf( stderr, "\rRound %3d, %d/%d", r, ++cnt, num_seq );
            fflush( stderr );
        }

        // step 1
        float* old_p = prices[t][i];
        float* new_p = prices[next_t][i];
        int* old_b = bidders[t][i];
        int* new_b = bidders[next_t][i];

        for( k = 0; k < num_bc; k++ ) {
            new_p[k] = old_p[k];
            new_b[k] = old_b[k];
        }

        for( j = 0; j < num_seq; j++ ) {
#if 0 // as long as the data is available in memory, why not use it all?
            if( (j == i) || !neighbor( i, j ) ) continue;
#endif
            for( k = 0; k < num_bc; k++ ) {
                if( ( prices[t][j][k] > new_p[k] ) 
                    || ( prices[t][j][k] == new_p[k] 
                         && bidders[t][j][k] > new_b[k] ) ) {
                    new_p[k] = prices[t][j][k];
                    new_b[k] = bidders[t][j][k];
                }
            }
        }


        k = alpha[t][i];
        if( values[i][locations[t][i]] < epsilon ) {
            new_b[k] = num_seq; // artificial nudge
        }

        // step 2
        if( old_p[k] <= new_p[k] && new_b[k] != i ) {
            // step 3
            #pragma omp atomic update
            (*num_swap)++;
            // 3.a update local values
            update_local_values( i );

            // 3.b choose a new assignment
            float best_g = -100.0;
            long long best_x = -1;
            long best_c = -1;
            val_map_it m;
            for( m = values[i].begin(); m != values[i].end(); m++ ) {
                long long x = m->first;
                long c = location_to_barcode( x );
                if( values[i][x] - new_p[c] > best_g ) {
                    best_g = values[i][x] - new_p[c];
                    best_x = x;
                    best_c = c;
                }
            }
            float next_g = -100.0;
            for( m = values[i].begin(); m != values[i].end(); m++ ) {
                long long x = m->first;
                long c = location_to_barcode( x );
                if( c == best_c ) continue;
                if( values[i][x] - new_p[c] > next_g ) {
                    next_g = values[i][x] - new_p[c];
                }
            }

            alpha[next_t][i] = best_c;
            locations[next_t][i] = best_x;
            new_b[best_c] = i;
            new_p[best_c] += ( best_g - next_g ) + epsilon;
        } else {
            // step 4
            alpha[next_t][i] = k;
            locations[next_t][i] = locations[t][i];
        }

        #pragma omp atomic update
        (*delta) += values[i][locations[next_t][i]] 
                    - prices[next_t][i][alpha[next_t][i]];
    } // for( i

    (*delta) /= num_seq;
}



int load_input( void )
{
    char* line = NULL;
    size_t len = 0;
    ssize_t r;
    do {
        r = getline( &line, &len, stdin );
        if( r < 0 ) break;
        int i, j, k;
        if( line[0] == '-' ) { // pre-selected barcode, put in exclusion list
            i = strcspn( line, BASES );
            j = strspn( line + i, BASES );
            if( j != cnf->bclen ) continue;
            line[i + j] = '\0';
            tabu.insert( bcseq_to_barcode( line+i ) );
            continue;
        }
        // targets
        i = strspn( line, BASES );
        j = strspn( line + i, SEPARATORS );
        if( j == 0 && i == 0 ) continue;
        k = 0;
        if( j ) {
            k = strspn( line + i + j, BASES );
        }
        line[i + j + k] = '\0';
        if( !is_legal( line ) ) {
            fprintf( stderr, "Warning, illegal sequence %s ignored.\n", line );
        } else {
            char* clean = strdup( line );
            if( j ) strcpy( clean + i, line + i + j );
            clean[i + k] = '\0';
            // add to the list
            nseqs++;
            seqs = (char**) realloc( seqs, nseqs * sizeof(char*) );
            seqs[nseqs - 1] = clean;
            hip = (int*) realloc( hip, nseqs * sizeof(int) );
            hip[nseqs - 1] = i;
        }
    } while( r >= 0 );

    free( line );
    return nseqs;
}


void eval_input( void )
{
    std::queue< std::string, std::list<std::string> > q_seq, q_copy;
    std::queue< int, std::list<int> > q_ofs;
    std::queue< double, std::list<double> > q_sc;
    
    char* line = NULL;
    size_t len = 0;
    ssize_t r;
    do {
        r = getline( &line, &len, stdin );
        if( r < 0 ) break;
        int i = strspn( line, BASES );
        int j = strspn( line + i, SEPARATORS );
        if( j == 0 ) continue;
        int k = strspn( line + i + j, BASES );
        if( k != strlen( cnf->hairpin_nts ) ) continue;
        char* clean = strdup( line );
        strcpy( clean + i, line + i + j );
        int l = strspn( line + i + j + k, SEPARATORS );
        int m = 0;
        if( l ) {
            m = strspn( line + i + j + k + l, BASES );
            strcpy( clean + i + k, line + i + j + k + l );
        }
        clean[i + k + m] = '\0';
        q_seq.push( clean );
        q_ofs.push( i );
        free( clean );
    } while( r >= 0 );
    free( line );

    size_t n = q_seq.size();
    if( verbose ) fprintf( stderr, "Evaluating %lu sequences...\n", n );
    unsigned int i;
    #pragma omp parallel for schedule(dynamic) ordered
    for( i = 0; i < n; i++ ) {
        char* xseq;
        int o;

        #pragma omp critical(queues)
        {
            xseq = strdup( q_seq.front().c_str() );
            q_seq.pop();
            o = q_ofs.front();
            q_ofs.pop();
        }

        double score = score_seq( 0, xseq, o );

        #pragma omp ordered
        {
            q_copy.push( xseq );
            q_sc.push( score );
        }
        free( xseq );
    }

    for( i = 0; i < n; i++ ) {
        fprintf( stdout, "%9.6f\t%s\n", q_sc.front(), q_copy.front().c_str() );
        q_sc.pop();
        q_copy.pop();
    }
        
}


void auction_barcodes( void )
{
    init();

    int j;
    long k;
    for( j = 0; j < num_seq; j++ ) {
        sequences[j] = seqs[j];
    }
    t = 0;

    // choose an "origin" (see populating the arrays, below)
    long long origin = 0;
    for( j = 0; j < cnf->bclen; j++ ) {
        // try to make it a 'good' one: CG/GC closing, and 60% CG/GC pairs
        int v = int_urn( 0, j == 0 || j == cnf->bclen-1 ? 1 : 9 ) % 4;
        origin <<= 4;
        origin |= v==0 ? 0x3 : v==1 ? 0xC : v==2 ? 0x6 : 0x9;
    }
    long org_bc = location_to_barcode( origin );
    
    // populate arrays
    for( j = 0; j < num_seq; j++ ) {
        for( k = 0; k < num_bc; k++ ) {
            prices[0][j][k] = 1.0 + drand48();
            bidders[0][j][k] = j;
        }
        // all agents start from the same location, and are artificially
        // outbid, which forces them to explore the neighborhood and start
        // spreading in all directions
        locations[0][j] = origin;
        alpha[0][j] = org_bc;
        bidders[0][j][alpha[0][j]] = num_seq;
    }

    // do the loop thing...
    int r = 0;
    int swaps;
    do {
        float delta;
        do_round( ++r, &swaps, &delta );
        if( verbose ) 
            fprintf( stderr, "\rRound %3d: %8d PFs computed, "
                             "%5d reassignments\n", r, num_scored, swaps );
        t++;
        t %= 2;
    // ... until no one complains anymore
    } while( swaps > 0 );

    // output
    fprintf( stderr, "\n" );
    for( j = 0; j < num_seq; j++ ) {
        long long a = locations[0][j];
        int len = 1 + strlen( sequences[j] ) + cnf->hp_tl;
        char* seq = (char*) malloc( len );
        sprintf( seq, "%s%.*s_%s_%s%s", cnf->tail5p_nts, 
                                        hip[j], sequences[j],
                                        cnf->hairpin_nts, 
                                        sequences[j]+hip[j],
                                        cnf->tail3p_nts );
        location_to_bcseq( a, seq + cnf->tl5p + hip[j] + 1 + cnf->hp5p,
                              seq + cnf->tl5p + hip[j] + 1 + cnf->hp3p );
        if( verbose ) fprintf( stderr, "%s\t%6.3f\n", seq, values[j][a] );
        fprintf( stdout, "%s\n", seq );
        free( seq );
    }

}


int main( int argc, char** argv )
{
    edo_args_info args_info;
    
    if( edo_cmdline_parser( argc, argv, &args_info ) != 0 )
        exit( 1 );
    
    if( args_info.verbose_given ) verbose = 1;

    if( args_info.lab_settings_given ) {
        if( !load_config( args_info.lab_settings_arg ) ) {
            exit( 1 );
        }
    }

    if( args_info.paramFile_given ) {
        load_fold_params( args_info.paramFile_arg );
    }

    if( args_info.eval_given ) {
        eval_input();
        exit( 0 );
    }

    edo_cmdline_parser_free( &args_info );

    load_input();
    if( nseqs == 0 ) {
        fprintf( stderr, "Not much work to do..." );
        return 0;
    }

    num_seq = nseqs;
    if( verbose )
        fprintf( stderr, "Loaded %d sequences, %ld barcode(s)"
                         " are reserved.\n\n", num_seq, tabu.size() );

    auction_barcodes();
    return 0;
}

